### ISSUE TITLE
# 关于单张图片读取训练的事情

### ISSUE THREAD

**Issue by Yulin-CC** [bubbliiiing/pspnet-tf2#1]
博主你好，我想问问关于 这份代码中 utils_fit.py 文件中的 
get_train_step_fn 函数，我的理解是这里是读取 batch_size 张图片，形成矩阵一起计算 loss 并平均
>- 我想实现的是数据集中不同图像用不同的 loss ，所以想到的是单张图像读取
>- 那如何单张图片读取并计算 loss 呢？

```python
def get_train_step_fn():
    @tf.function
    def train_step(images, labels, net, optimizer, loss, aux_branch, f_score, iou_score):
        with tf.GradientTape() as tape:
            prediction = net(images, training=True)     # 网络的输出 ， 维度是 (x, 473, 473, 21)，tensor 类型
            if aux_branch:                                           
                 aux_loss    = loss(labels[0], prediction[0])
                 main_loss   = loss(labels[1], prediction[1])
                 loss_value  = aux_loss + main_loss
             else:
                 loss_value  = loss(labels, prediction)    # 博主博主看这里看这里啊看这里

        # 计算梯度和优化
        grads = tape.gradient(loss_value, net.trainable_variables)
        optimizer.apply_gradients(zip(grads, net.trainable_variables))
        
        if aux_branch:          # false
            _f_score  = tf.reduce_mean(f_score(labels[1], prediction[1]))
        else:
            _f_score  = tf.reduce_mean(f_score(labels, prediction))
            _iou_score = tf.reduce_mean(iou_score(labels, prediction))
        return loss_value, _f_score, _iou_score
    return train_step
```

我修改成了这样子
>- 确实也能进行训练，但是验证集的表现并没有随之提高，就好像，网络没有更新
```python
def get_train_step_fn():
    @tf.function
    def train_step(images, labels, tabs, net, optimizer, loss, aux_branch, f_score, iou_score):
        with tf.GradientTape() as tape:
            predictions = net(images, training=True)  # 网络的输出 ， 维度是 (x, 473, 473, 21)，tensor 类型9

            # 损失值
            Loss = []

            for i in range(len(images)):
   
                image  = tf.reshape(images[i], [1, 473, 473, 3])               # 博主博主看这里看这里啊看这里
                label  = tf.reshape(labels[i], [1, 473, 473, 3])

                prediction = net(image, training=True)  # 网络的输出 ， 维度是 (x, 473, 473, 21)，tensor 类型9
                lce_loss = loss(label, prediction)  # CE_loss，输出的 loss 是一个 tensor 值, 每一张图片的损失，接下来要求 4 张图片的平均值
                Loss.append(lce_loss)
                loss_value = tf.reduce_mean(Loss)

        # 计算梯度和优化模型
        grads = tape.gradient(loss_value, net.trainable_variables)
        optimizer.apply_gradients(zip(grads, net.trainable_variables))

        if aux_branch:          # false
            _f_score  = tf.reduce_mean(f_score(labels[1], prediction[1]))
        else:
            _f_score  = tf.reduce_mean(f_score(labels, predictions))
            _iou_score = tf.reduce_mean(iou_score(labels, predictions))
        return lce_loss, _f_score, _iou_score
    return train_step
```

打扰小哥了！！

**Comment by bubbliiiing** [comment]
不可单张图像传入，bn层会出错
